File: js/modules/gemini/client/modelRegistry.js
/**
 * Browser-side Gemini model policy.
 *
 * Preview model ids expire. All persisted or user-controlled model values must
 * pass through this registry before they are shown or sent to the backend.
 */
(function () {
    'use strict';

    const DEFAULTS = Object.freeze({
        live: 'gemini-3.1-flash-live-preview',
        textBrain: 'gemini-3.5-flash-lite',
        transcription: 'gemini-3.6-flash'
    });

    const MODELS = Object.freeze({
        live: Object.freeze([
            Object.freeze({
                id: 'gemini-3.1-flash-live-preview',
                label: 'Gemini 3.1 Flash Live Preview (recommended)',
                summary: 'Current low-latency Live audio model with native input and output transcription.',
                capabilities: Object.freeze({
                    audioOutput: true,
                    inputAudioTranscription: true,
                    outputAudioTranscription: true,
                    affectiveDialog: false,
                    proactiveAudio: false
                })
            }),
            Object.freeze({
                id: 'gemini-2.5-flash-native-audio-preview-12-2025',
                label: 'Gemini 2.5 Flash Native Audio (12-2025 fallback)',
                summary: 'Compatibility fallback with native transcription, affective dialog, and proactive audio.',
                capabilities: Object.freeze({
                    audioOutput: true,
                    inputAudioTranscription: true,
                    outputAudioTranscription: true,
                    affectiveDialog: true,
                    proactiveAudio: true
                })
            })
        ]),
        textBrain: Object.freeze([
            Object.freeze({
                id: 'gemini-3.5-flash-lite',
                label: 'Gemini 3.5 Flash-Lite (recommended)',
                summary: 'Fast, large-context extraction model for Mode 2 and EveOS context relay.',
                capabilities: Object.freeze({ textGeneration: true, largeContext: true })
            }),
            Object.freeze({
                id: 'gemini-3.5-flash',
                label: 'Gemini 3.5 Flash',
                summary: 'Higher-capability text-brain option with a larger quota cost.',
                capabilities: Object.freeze({ textGeneration: true, largeContext: true })
            }),
            Object.freeze({
                id: 'gemini-3.6-flash',
                label: 'Gemini 3.6 Flash',
                summary: 'Newest general Flash option for heavier extraction and reasoning turns.',
                capabilities: Object.freeze({ textGeneration: true, largeContext: true })
            }),
            Object.freeze({
                id: 'gemini-3.1-flash-lite',
                label: 'Gemini 3.1 Flash-Lite (compatibility fallback)',
                summary: 'Conservative fallback when newer text-brain models are unavailable.',
                capabilities: Object.freeze({ textGeneration: true, largeContext: true })
            })
        ]),
        transcription: Object.freeze([
            Object.freeze({
                id: 'gemini-3.6-flash',
                label: 'Gemini 3.6 Flash',
                summary: 'Fallback audio-file transcription model.',
                capabilities: Object.freeze({ audioInput: true, textGeneration: true })
            }),
            Object.freeze({
                id: 'gemini-3.5-flash',
                label: 'Gemini 3.5 Flash',
                summary: 'Compatibility fallback for audio-file transcription.',
                capabilities: Object.freeze({ audioInput: true, textGeneration: true })
            })
        ])
    });

    const MIGRATIONS = Object.freeze({
        live: Object.freeze({
            'gemini-2.5-flash-native-audio-latest': DEFAULTS.live,
            'gemini-2.5-flash-preview-native-audio-dialog': DEFAULTS.live,
            'gemini-2.5-flash-experimental-native-audio-thinking-dialog': DEFAULTS.live,
            'gemini-2.0-flash-live-001': DEFAULTS.live,
            'gemini-live-2.5-flash-preview': DEFAULTS.live
        }),
        textBrain: Object.freeze({
            'gemini-2.5-flash-lite': DEFAULTS.textBrain,
            'gemini-2.5-flash': 'gemini-3.5-flash',
            'gemini-2.5-pro': 'gemini-3.6-flash',
            'gemini-2.0-flash-lite': DEFAULTS.textBrain,
            'gemini-2.0-flash': 'gemini-3.5-flash'
        }),
        transcription: Object.freeze({
            'gemini-2.0-flash': DEFAULTS.transcription,
            'gemini-2.5-flash': DEFAULTS.transcription
        })
    });

    const STORAGE_BINDINGS = Object.freeze({
        selectedModel: 'live',
        textBrainModel: 'textBrain'
    });

    function modelList(kind) {
        return MODELS[kind] || [];
    }

    function resolve(kind, value) {
        const candidate = String(value || '').trim();
        const migrated = MIGRATIONS[kind]?.[candidate] || candidate;
        return modelList(kind).some(function (model) { return model.id === migrated; })
            ? migrated
            : DEFAULTS[kind];
    }

    function getModel(kind, value) {
        const id = resolve(kind, value);
        return modelList(kind).find(function (model) { return model.id === id; }) || null;
    }

    function getCapabilities(kind, value) {
        return Object.assign({}, getModel(kind, value)?.capabilities || {});
    }

    function getFallback(kind, value) {
        const models = modelList(kind);
        const selected = resolve(kind, value);
        const selectedIndex = models.findIndex(function (model) { return model.id === selected; });
        return selectedIndex >= 0 && selectedIndex + 1 < models.length
            ? models[selectedIndex + 1].id
            : '';
    }

    function migrateStorage(storage) {
        const target = storage || window.localStorage;
        const changes = [];
        Object.keys(STORAGE_BINDINGS).forEach(function (storageKey) {
            const kind = STORAGE_BINDINGS[storageKey];
            let previous = '';
            try { previous = String(target?.getItem?.(storageKey) || '').trim(); }
            catch (error) { return; }
            const next = resolve(kind, previous);
            if (previous === next) return;
            try { target?.setItem?.(storageKey, next); }
            catch (error) { return; }
            changes.push({ storageKey: storageKey, kind: kind, from: previous, to: next });
        });
        if (changes.length && typeof window.dispatchEvent === 'function' && typeof window.CustomEvent === 'function') {
            window.dispatchEvent(new window.CustomEvent('eve:gemini-models-migrated', { detail: { changes: changes } }));
        }
        return changes;
    }

    function populateSelect(select, kind, value) {
        if (!select) return resolve(kind, value);
        const selected = resolve(kind, value);
        select.replaceChildren();
        modelList(kind).forEach(function (model) {
            const option = document.createElement('option');
            option.value = model.id;
            option.textContent = model.label;
            option.selected = model.id === selected;
            select.appendChild(option);
        });
        select.value = selected;
        return selected;
    }

    window.EveGeminiModelRegistry = Object.freeze({
        defaults: DEFAULTS,
        models: MODELS,
        migrations: MIGRATIONS,
        resolve: resolve,
        getModel: getModel,
        getModels: function (kind) { return modelList(kind).slice(); },
        getFallback: getFallback,
        getCapabilities: getCapabilities,
        migrateStorage: migrateStorage,
        populateSelect: populateSelect
    });

    migrateStorage();
})();

File: server/gemini-backend/interactions/main_server_files/api_configuration/model_registry.py
"""Supported Gemini model registry and persisted-id migration helpers.

Keep model policy in one place. Preview aliases are short-lived, so every caller
must resolve user-controlled or persisted values through this module before an
API request is made.
"""

from __future__ import annotations

from copy import deepcopy


LIVE_DEFAULT_MODEL = "gemini-3.1-flash-live-preview"
TEXT_BRAIN_DEFAULT_MODEL = "gemini-3.5-flash-lite"
TRANSCRIPTION_DEFAULT_MODEL = "gemini-3.6-flash"


LIVE_MODELS = {
    LIVE_DEFAULT_MODEL: {
        "label": "Gemini 3.1 Flash Live Preview",
        "capabilities": {
            "audio_output": True,
            "input_audio_transcription": True,
            "output_audio_transcription": True,
            "affective_dialog": False,
            "proactive_audio": False,
        },
    },
    "gemini-2.5-flash-native-audio-preview-12-2025": {
        "label": "Gemini 2.5 Flash Native Audio (12-2025 fallback)",
        "capabilities": {
            "audio_output": True,
            "input_audio_transcription": True,
            "output_audio_transcription": True,
            "affective_dialog": True,
            "proactive_audio": True,
        },
    },
}


TEXT_BRAIN_MODELS = {
    "gemini-3.5-flash-lite": {
        "label": "Gemini 3.5 Flash-Lite (recommended)",
        "capabilities": {"text_generation": True, "large_context": True},
    },
    "gemini-3.5-flash": {
        "label": "Gemini 3.5 Flash",
        "capabilities": {"text_generation": True, "large_context": True},
    },
    "gemini-3.6-flash": {
        "label": "Gemini 3.6 Flash",
        "capabilities": {"text_generation": True, "large_context": True},
    },
    "gemini-3.1-flash-lite": {
        "label": "Gemini 3.1 Flash-Lite (compatibility fallback)",
        "capabilities": {"text_generation": True, "large_context": True},
    },
}


TRANSCRIPTION_MODELS = {
    TRANSCRIPTION_DEFAULT_MODEL: {
        "label": "Gemini 3.6 Flash",
        "capabilities": {"audio_input": True, "text_generation": True},
    },
    "gemini-3.5-flash": {
        "label": "Gemini 3.5 Flash",
        "capabilities": {"audio_input": True, "text_generation": True},
    },
}


LEGACY_MODEL_MIGRATIONS = {
    "live": {
        "gemini-2.5-flash-native-audio-latest": LIVE_DEFAULT_MODEL,
        "gemini-2.5-flash-preview-native-audio-dialog": LIVE_DEFAULT_MODEL,
        "gemini-2.5-flash-experimental-native-audio-thinking-dialog": LIVE_DEFAULT_MODEL,
        "gemini-2.0-flash-live-001": LIVE_DEFAULT_MODEL,
        "gemini-live-2.5-flash-preview": LIVE_DEFAULT_MODEL,
    },
    "text_brain": {
        "gemini-2.5-flash-lite": TEXT_BRAIN_DEFAULT_MODEL,
        "gemini-2.5-flash": "gemini-3.5-flash",
        "gemini-2.5-pro": "gemini-3.6-flash",
        "gemini-2.0-flash-lite": TEXT_BRAIN_DEFAULT_MODEL,
        "gemini-2.0-flash": "gemini-3.5-flash",
    },
    "transcription": {
        "gemini-2.0-flash": TRANSCRIPTION_DEFAULT_MODEL,
        "gemini-2.5-flash": TRANSCRIPTION_DEFAULT_MODEL,
    },
}


_REGISTRIES = {
    "live": LIVE_MODELS,
    "text_brain": TEXT_BRAIN_MODELS,
    "transcription": TRANSCRIPTION_MODELS,
}

_DEFAULTS = {
    "live": LIVE_DEFAULT_MODEL,
    "text_brain": TEXT_BRAIN_DEFAULT_MODEL,
    "transcription": TRANSCRIPTION_DEFAULT_MODEL,
}


def resolve_model(kind: str, name: object) -> str:
    """Return a supported model id, migrating known aliases and rejecting unknown ids."""
    if kind not in _REGISTRIES:
        raise ValueError(f"Unknown Gemini model kind: {kind}")
    candidate = str(name or "").strip()
    candidate = LEGACY_MODEL_MIGRATIONS.get(kind, {}).get(candidate, candidate)
    return candidate if candidate in _REGISTRIES[kind] else _DEFAULTS[kind]


def resolve_live_model(name: object) -> str:
    return resolve_model("live", name)


def resolve_text_brain_model(name: object) -> str:
    return resolve_model("text_brain", name)


def resolve_transcription_model(name: object) -> str:
    return resolve_model("transcription", name)


def model_capabilities(kind: str, name: object) -> dict:
    resolved = resolve_model(kind, name)
    return deepcopy(_REGISTRIES[kind][resolved]["capabilities"])


def model_options(kind: str) -> tuple[str, ...]:
    if kind not in _REGISTRIES:
        raise ValueError(f"Unknown Gemini model kind: {kind}")
    return tuple(_REGISTRIES[kind].keys())


def public_registry() -> dict:
    """Serializable registry shape used by diagnostics and parity smokes."""
    return {
        "defaults": dict(_DEFAULTS),
        "models": deepcopy(_REGISTRIES),
        "migrations": deepcopy(LEGACY_MODEL_MIGRATIONS),
    }
